In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("EV_Adoption_and_Range_Anxiety_Dataset.csv")

In [3]:
df.head()

,Buyer_ID,Age,Gender,Annual_Income_USD,City_Type,Daily_Commute_km,Number_of_Cars_Owned,Current_Car_Type,Charging_Stations_Near_Home,Charging_Stations_Near_Work,Home_Charging_Possible,Environmental_Concern_Level,Subsidy_Available,Range_Anxiety_Level,Will_Buy_EV
0,EV00001,56,Male,44906.0,Rural,18.1,1,Truck,1,0,Yes,4.0,Yes,Low,No
1,EV00002,36,Male,88754.0,Rural,5.0,1,SUV,2,3,Yes,1.0,Yes,Low,No
2,EV00003,42,Male,155958.0,Urban,32.2,2,Sedan,14,3,No,2.0,Yes,Medium,Yes
3,EV00004,31,Male,37059.0,Rural,89.9,1,Sedan,2,1,Yes,4.0,No,Low,No
4,EV00005,48,Female,48552.0,Rural,71.0,1,Sedan,2,0,Yes,3.0,Yes,Low,No


In [4]:
df["Will_Buy_EV"].value_counts(dropna=False)

Will_Buy_EV
No     8250
Yes    1750
Name: count, dtype: int64

In [5]:
df.isna().sum()

Buyer_ID                         0
Age                              0
Gender                           0
Annual_Income_USD              178
City_Type                        0
Daily_Commute_km               181
Number_of_Cars_Owned             0
Current_Car_Type                 0
Charging_Stations_Near_Home      0
Charging_Stations_Near_Work      0
Home_Charging_Possible           0
Environmental_Concern_Level    184
Subsidy_Available                0
Range_Anxiety_Level              0
Will_Buy_EV                      0
dtype: int64

In [6]:
df["Annual_Income_USD"] = df["Annual_Income_USD"].fillna(df["Annual_Income_USD"].median())

In [7]:
df["Daily_Commute_km"] = df["Daily_Commute_km"].fillna(df["Daily_Commute_km"].median())

In [8]:
df["Environmental_Concern_Level"] = df["Environmental_Concern_Level"].fillna(df["Environmental_Concern_Level"].mean())

In [9]:
df.isna().sum()

Buyer_ID                       0
Age                            0
Gender                         0
Annual_Income_USD              0
City_Type                      0
Daily_Commute_km               0
Number_of_Cars_Owned           0
Current_Car_Type               0
Charging_Stations_Near_Home    0
Charging_Stations_Near_Work    0
Home_Charging_Possible         0
Environmental_Concern_Level    0
Subsidy_Available              0
Range_Anxiety_Level            0
Will_Buy_EV                    0
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Buyer_ID                     10000 non-null  object 
 1   Age                          10000 non-null  int64  
 2   Gender                       10000 non-null  object 
 3   Annual_Income_USD            10000 non-null  float64
 4   City_Type                    10000 non-null  object 
 5   Daily_Commute_km             10000 non-null  float64
 6   Number_of_Cars_Owned         10000 non-null  int64  
 7   Current_Car_Type             10000 non-null  object 
 8   Charging_Stations_Near_Home  10000 non-null  int64  
 9   Charging_Stations_Near_Work  10000 non-null  int64  
 10  Home_Charging_Possible       10000 non-null  object 
 11  Environmental_Concern_Level  10000 non-null  float64
 12  Subsidy_Available            10000 non-null  object 
 13  Range_Anxiety_Lev

In [12]:
df['Will_Buy_EV'] = df['Will_Buy_EV'].map({'Yes': 1, 'No': 0})
df.head()

,Buyer_ID,Age,Gender,Annual_Income_USD,City_Type,Daily_Commute_km,Number_of_Cars_Owned,Current_Car_Type,Charging_Stations_Near_Home,Charging_Stations_Near_Work,Home_Charging_Possible,Environmental_Concern_Level,Subsidy_Available,Range_Anxiety_Level,Will_Buy_EV
0,EV00001,56,Male,44906.0,Rural,18.1,1,Truck,1,0,Yes,4.0,Yes,Low,0
1,EV00002,36,Male,88754.0,Rural,5.0,1,SUV,2,3,Yes,1.0,Yes,Low,0
2,EV00003,42,Male,155958.0,Urban,32.2,2,Sedan,14,3,No,2.0,Yes,Medium,1
3,EV00004,31,Male,37059.0,Rural,89.9,1,Sedan,2,1,Yes,4.0,No,Low,0
4,EV00005,48,Female,48552.0,Rural,71.0,1,Sedan,2,0,Yes,3.0,Yes,Low,0


In [13]:
x = df.drop(["Buyer_ID","Will_Buy_EV"],axis=1)
y = df["Will_Buy_EV"]

In [14]:
category_cols = [
    "Gender",
    "Daily_Commute_km",
    "Current_Car_Type",
    "Home_Charging_Possible",
    "Subsidy_Available",
    "Range_Anxiety_Level",
    "City_Type"
]

In [15]:
x = pd.get_dummies(x,columns=category_cols,drop_first=True,dtype=int)

In [16]:
x

,Age,Annual_Income_USD,Number_of_Cars_Owned,Charging_Stations_Near_Home,Charging_Stations_Near_Work,Environmental_Concern_Level,Gender_Male,Gender_Other,Daily_Commute_km_5.1,Daily_Commute_km_5.2,...,Daily_Commute_km_135.5,Current_Car_Type_SUV,Current_Car_Type_Sedan,Current_Car_Type_Truck,Home_Charging_Possible_Yes,Subsidy_Available_Yes,Range_Anxiety_Level_Low,Range_Anxiety_Level_Medium,City_Type_Suburban,City_Type_Urban
0,56,44906.0,1,1,0,4.0,1,0,0,0,...,0,0,0,1,1,1,1,0,0,0
1,36,88754.0,1,2,3,1.0,1,0,0,0,...,0,1,0,0,1,1,1,0,0,0
2,42,155958.0,2,14,3,2.0,1,0,0,0,...,0,0,1,0,0,1,0,1,0,1
3,31,37059.0,1,2,1,4.0,1,0,0,0,...,0,0,1,0,1,0,1,0,0,0
4,48,48552.0,1,2,0,3.0,0,0,0,0,...,0,0,1,0,1,1,1,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,50,61279.0,1,8,3,3.0,1,0,0,0,...,0,0,1,0,0,1,0,1,0,1
9996,34,108208.0,2,9,7,5.0,1,0,0,0,...,0,0,0,0,0,0,1,0,0,1
9997,60,97743.0,2,10,3,5.0,0,0,0,0,...,0,0,1,0,0,1,1,0,0,1
9998,31,103628.0,2,10,10,1.0,0,0,0,0,...,0,1,0,0,1,1,1,0,0,1


In [17]:
y

0       0
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    0
9998    0
9999    0
Name: Will_Buy_EV, Length: 10000, dtype: int64

In [18]:
from sklearn.model_selection import train_test_split

In [19]:
x_train, x_test, y_train, y_test = train_test_split(
    x,y,test_size=0.20,stratify=y, random_state = 42
)

In [21]:
from sklearn.preprocessing import StandardScaler

In [22]:
scaler = StandardScaler()

In [23]:
scaler

StandardScaler()

In [24]:
x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)

In [25]:
from sklearn.linear_model import LogisticRegression

In [26]:
model = LogisticRegression(
    max_iter = 1000,
    class_weight = "balanced"
)

In [27]:
model

LogisticRegression(class_weight='balanced', max_iter=1000)

In [28]:
model.fit(x_train_scaled,y_train)

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:451: OptimizeWarning: Unknown solver options: iprint
  opt_res = optimize.minimize(


LogisticRegression(class_weight='balanced', max_iter=1000)

In [29]:
y_pred = model.predict(x_test_scaled)

In [30]:
y_pred

array([1, 0, 1, ..., 1, 0, 1], shape=(2000,))

In [31]:
compariosn = pd.DataFrame({
    "Actual": y_test,
    "Prediction" : y_pred
})

In [32]:
compariosn

,Actual,Prediction
8799,0,1
9510,0,0
5917,0,1
8995,0,0
5256,0,1
...,...,...
8921,0,0
2241,1,1
6351,0,1
4041,0,0


In [33]:
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score

In [34]:
accuracy = accuracy_score(y_test,y_pred)*100
precision = precision_score(y_test,y_pred)*100
recall =  recall_score(y_test,y_pred)*100
f1 = f1_score(y_test,y_pred)*100

In [37]:
print(f"accuracy: {accuracy:.2f}%")
print(f"precision:{precision:.2f}%")
print(f"recall:{recall:.2f}%")
print(f"f1:{f1_score:.2f}%")

accuracy: 79.50%
precision:44.19%
recall:65.14%
f1:52.66%


In [38]:
from sklearn.metrics import confusion_matrix

In [39]:
cm = confusion_matrix(y_test,y_pred)

In [40]:
cm

array([[1362,  288],
       [ 122,  228]])

In [41]:
from sklearn.metrics import classification_report

In [42]:
classification = classification_report(y_test,y_pred)

In [43]:
print("classification :\n",classification)

classification :
               precision    recall  f1-score   support

           0       0.92      0.83      0.87      1650
           1       0.44      0.65      0.53       350

    accuracy                           0.80      2000
   macro avg       0.68      0.74      0.70      2000
weighted avg       0.83      0.80      0.81      2000



In [44]:
model = LogisticRegression(
    class_weight='balanced',
    max_iter=1000
)

In [45]:
model.fit(x_train, y_train)

y_pred = model.predict(x_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:451: OptimizeWarning: Unknown solver options: iprint
  opt_res = optimize.minimize(


Accuracy: 0.8025
[[1298  352]
 [  43  307]]
              precision    recall  f1-score   support

           0       0.97      0.79      0.87      1650
           1       0.47      0.88      0.61       350

    accuracy                           0.80      2000
   macro avg       0.72      0.83      0.74      2000
weighted avg       0.88      0.80      0.82      2000



C:\ProgramData\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Final Model Conclusion

The final **Logistic Regression model with `class_weight="balanced"`** was selected as the final model for predicting the target variable, where **0 = No** and **1 = Yes**.

The model achieved an overall **accuracy of 80.25%**. For the positive class (**Yes**), the model achieved **47% precision, 88% recall, and 61% F1-score**. The use of `class_weight="balanced"` significantly improved the detection of positive cases, increasing Class 1 recall from **65% to 88%** and F1-score from **53% to 61%**.

The model correctly identified **307 out of 350 actual Yes cases**, while 43 Yes cases were classified as No. Although the model produced some false positives, the improved recall indicates that it is effective at identifying positive cases.

Therefore, the **Logistic Regression model with balanced class weights** was selected as the final model because it provides a better balance between overall performance and detection of the minority **Yes** class.

**Model Evaluation**<br>
       ↓<br>
**Confusion Matrix**<br>
       ↓<br>
**Classification Report**<br>
       ↓<br>
**Final Model Conclusion**